src/Utility/ExprFunctions.C

```diff
@@ -223,24 +271,142 @@ Real EvalFunction::evaluate (const Vec3& X) const
 }
 
 
+Real EvalFunction::deriv (const Vec3& X, int dir) const
+{
+  if (dir < 1 || dir > 3 || !gradient[--dir])
+    return Real(0);
+
+  return gradient[dir]->evaluate(X);
+}
+
+
+Real EvalFunction::dderiv (const Vec3& X, int d1, int d2) const
+{
+  if (d1 > d2)
+    std::swap(d1,d2); // Assuming symmetry
+
+  if (d1 < 1 || d2 > 3)
+    return Real(0);
+
+  // Assuming Voigt notation ordering; 11, 22, 33, 12, 23, 13
+  if (d2-d1 == 0) // diagoal term, 11, 22 and 33
+    --d1;
+  else if (d2-d1 == 1) // off-diagonal term, 12 and 23
+    d1 = d2+1;
+  else // off-diagonal term, 13
+    d1 = 5;
+
+  return dgradient[d1] ? dgradient[d1]->evaluate(X) : Real(0);
+}
+
+
+/*!
+  \brief Static helper that splits a function expression into components.
+*/
+
+static std::vector<std::string> splitComps (const std::string& functions,
+                                            const std::string& variables)
+{
+  std::vector<std::string> comps;
+  size_t pos1 = functions.find("|");
+  size_t pos2 = 0;
+  while (pos2 < functions.size())
+  {
+    std::string func(variables);
+    if (!func.empty() && func[func.size()-1] != ';')
+      func += ';';
+    if (pos1 == std::string::npos)
+      func += functions.substr(pos2);
+    else
+      func += functions.substr(pos2,pos1-pos2);
+    comps.push_back(func);
+    pos2 = pos1 > 0 && pos1 < std::string::npos ? pos1+1 : pos1;
+    pos1 = functions.find("|",pos1+1);
+  }
+
+  return comps;
+}
+
+
+EvalFunctions::EvalFunctions (const std::string& functions,
+                              const std::string& variables)
+{
+  std::vector<std::string> components = splitComps(functions,variables);
+  for (const auto& comp : components)
+    p.push_back(new EvalFunction(comp.c_str()));
+}
+
+
+EvalFunctions::~EvalFunctions ()
+{
+  for (auto& it : p)
+    delete it;
+}
+
+
+void EvalFunctions::addDerivative (const std::string& functions,
+                                   const std::string& variables, int d1, int d2)
+{
+  std::vector<std::string> components = splitComps(functions,variables);
+  for (size_t i = 0; i < p.size() && i < components.size(); i++)
+    p[i]->addDerivative(components[i],variables,d1,d2);
+}
+
+
 template<>
 Vec3 VecFuncExpr::evaluate (const Vec3& X) const
 {
   Vec3 result;
-  for (size_t i = 0; i < 3 && i < p.size(); ++i)
+
+  for (size_t i = 0; i < 3 && i < nsd; ++i)
     result[i] = (*p[i])(X);
 
   return result;
 }
 
 
 template<>
+Vec3 VecFuncExpr::deriv (const Vec3& X, int dir) const
+{
+  Vec3 result;
+
+  for (size_t i = 0; i < 3 && i < nsd; ++i)
```

Comment:
You mean `for (size_t i = 0; i < 3 && i < nsd && p[i]; ++i)` ?

Not sure that is necessary, because nsd is set to p.size() in the VexFuncExpr constructor (via the default setNoDim()) method, and should only contain non-nulls.